README.md
# 介绍
-----

[![npm version](https://badge.fury.io/js/think_apollo.svg)](https://badge.fury.io/js/think_apollo)

Apollo Plugin for Koatty.

# 安装
-----

```
npm i koatty_apollo
```

# 使用
-----

## Koatty

1、项目中增加plugin

```shell
koatty plugin apollo;
```

2、修改 plugin/ApolloPlugin.ts:

```
import { Koatty, Plugin, IPlugin } from "koatty";
import Apollo from "koatty_apollo";

@Plugin()
export class ApolloPlugin implements IPlugin {
    run(options: any, app: Koatty) {
        return Apollo(options, app);
    }
}
```

3、项目plugin配置 config/plugin.ts:
```
list: ["ApolloPlugin"], //加载的插件列表,执行顺序按照数组元素顺序
    config: { //插件配置
    ...,
    ApolloPlugin: {
        host: 'http://127.0.0.1:8080', //your-config-server-url
        appId: 'test', //your-config-appId
        clusterName: 'default', //cluster-name
        namespace: 'application', //your-config-namespace
        enableUpdateNotification: true, // boolean=true set to false to disable update notification.
        enableFetch: true, //boolean=false set to true to enable the feature
        fetchTimeout: 30000, // number=30000 timeout in milliseconds before the initial fetch or interval fetch result in an FETCH_TIMEOUT error.
        fetchInterval: 3 * 60 * 1000, // number = 3 * 60 * 1000 interval in milliseconds to pull the new configurations.Defaults to 5 minutes.Setting this option to 0 will disable the feature.
        fetchCachedConfig: true, // boolean = true whether refresh configurations by fetching the restful API with caches.Defaults to true.If you want to always fetch the latest configurations(not recommend), set the option to false
    }
}


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.1.7](https://github.com/Koatty/koatty_apollo/compare/v2.1.6...v2.1.7) (2023-01-13)


### Bug Fixes

* audit ([ead59ef](https://github.com/Koatty/koatty_apollo/commit/ead59ef0a690d02140d8ee4ec3d40b7157402688))
* upgrade deps ([3264277](https://github.com/Koatty/koatty_apollo/commit/3264277c01d97573fa78f2aff3764eed67155204))

### [2.1.6](https://github.com/Koatty/koatty_apollo/compare/v2.1.4...v2.1.6) (2022-09-05)

### [2.1.5](https://github.com/Koatty/koatty_apollo/compare/v2.1.4...v2.1.5) (2022-05-27)

### 2.1.4 (2021-11-20)


index.js
/**
 *
 * @author     richen
 * @copyright  Copyright (c) 2017 - <richenlin(at)gmail.com>
 * @license    MIT
 * @version    17/6/12
 */
const apollo = require('ctrip-apollo')
const lib = require("think_lib");
const logger = require("think_logger");

/**
 * default options
 */
const defaultOptions = {
    host: 'http://127.0.0.1:8080', //your-config-server-url
    appId: 'test', //your-config-appId
    clusterName: 'default', //cluster-name
    namespace: 'application', //your-config-namespace
    enableUpdateNotification: true, // boolean=true set to false to disable update notification.
    enableFetch: true, //boolean=false set to true to enable the feature
    fetchTimeout: 30000, // number=30000 timeout in milliseconds before the initial fetch or interval fetch result in an FETCH_TIMEOUT error.
    fetchInterval: 3 * 60 * 1000, // number = 3 * 60 * 1000 interval in milliseconds to pull the new configurations.Defaults to 5 minutes.Setting this option to 0 will disable the feature.
    fetchCachedConfig: true, // boolean = true whether refresh configurations by fetching the restful API with caches.Defaults to true.If you want to always fetch the latest configurations(not recommend), set the option to false
};

module.exports = async function (options, app) {
    options = options ? lib.extend(defaultOptions, options, true) : defaultOptions;

    /**
     * 
     *
     * @param {*} newConfig
     * @returns
     */
    const reFreshConfig = function (newConfig) {
        if (lib.isEmpty(newConfig)) {
            return;
        }
        try {
            let typeConfig = {};
            for (const n in newConfig) {
                if (n.includes(".")) {
                    let type = n.slice(0, n.indexOf("."));
                    let key = n.slice(n.indexOf(".") + 1);
                    if (type && key) {
                        !typeConfig[type] && (typeConfig[type] = {});
                        if (lib.isJSONStr(newConfig[n])) {
                            typeConfig[type][key] = JSON.parse(newConfig[n]);
                        } else {
                            typeConfig[type][key] = newConfig[n];
                        }
                    }
                } else {
                    !typeConfig["config"] && (typeConfig["config"] = {});
                    if (lib.isJSONStr(newConfig[n])) {
                        typeConfig["config"][n] = JSON.parse(newConfig[n]);
                    } else {
                        typeConfig["config"][n] = newConfig[n];
                    }
                }
            }
            let appConfigs = {};
            if (app.setMap) {
                appConfigs = app.getMap("configs") || {};
                appConfigs = lib.extend(appConfigs, typeConfig, true);
                app.setMap("configs", appConfigs);
            } else if (app._caches.configs) {
                appConfigs = app._caches.configs || {};
                appConfigs = lib.extend(appConfigs, typeConfig, true);
                app._caches.configs = appConfigs;
            }
        } catch (err) {
            logger.error(err.stack || err);
        }
    }

    /**
     *
     *
     * @param {*} opt
     * @param {*} app
     */
    const initApolo = async function (opt, app) {
        // Instantiate Apollo
        const apolloClient = apollo(opt).namespace(opt.namespace, 'JSON');

        if (opt.enableUpdateNotification) {
            apolloClient.on('change', ({
                key,
                oldValue,
                newValue
            }) => {
                logger.info(`Apollo configuration key ${key} has been refreshed.`);
                reFreshConfig({ [key]: newValue });
            });
        }

        const ns = await apolloClient.ready().catch(err => {
            return Promise.reject(err);
        });
        logger.info('Apollo configuration center initialization is complete.');
        const apolloConfigs = await ns.config() || {};
        if (!lib.isEmpty(apolloConfigs)) {
            reFreshConfig(apolloConfigs);
        }

        return apolloClient;
    }
    // app.once('appReady', async () => {
    //     const client = await initApolo(options, app);
    //     lib.define(app, "apolloClient", client, true);
    // });

    const client = await initApolo(options, app);
    lib.define(app, "apolloClient", client, true);

    return async function (ctx, next) {
        if (!app.apolloClient || !app.apolloClient._ready) {
            const client = await initApolo(options, app);
            lib.define(app, "apolloClient", client, true);
        }
        return next();
    };
};

src/index.ts
/*
 * @Author: richen
 * @Date: 2020-07-27 11:32:15
 * @LastEditTime: 2020-12-24 15:28:13
 * @Description:
 * @Copyright (c) - <richenlin(at)gmail.com>
 */
import * as helper from 'koatty_lib';
import { DefaultLogger as logger } from 'koatty_logger';
const apollo = require("ctrip-apollo");

interface PluginOptions {
    host: string;
    appId: string;
    clusterName: string;
    namespace: string;
    enableUpdateNotification: boolean;
    enableFetch: boolean;
    fetchTimeout: number;
    fetchInterval: number;
    fetchCachedConfig: boolean;
}

interface UpdateNotification {
    key: string;
    oldValue: any;
    newValue: any;
}

/**
 * default options
 */
const defaultOptions: PluginOptions = {
    host: 'http://127.0.0.1:8080', //your-config-server-url
    appId: 'test', //your-config-appId
    clusterName: 'default', //cluster-name
    namespace: 'application', //your-config-namespace
    enableUpdateNotification: true, // boolean=true set to false to disable update notification.
    enableFetch: true, //boolean=false set to true to enable the feature
    fetchTimeout: 30000, // number=30000 timeout in milliseconds before the initial fetch or interval fetch result in an FETCH_TIMEOUT error.
    fetchInterval: 3 * 60 * 1000, // number = 3 * 60 * 1000 interval in milliseconds to pull the new configurations.Defaults to 5 minutes.Setting this option to 0 will disable the feature.
    fetchCachedConfig: true // boolean = true whether refresh configurations by fetching the restful API with caches.Defaults to true.If you want to always fetch the latest configurations(not recommend), set the option to false
};

/**
 * Wait for a period of time (ms)
 *
 * @param {number} ms
 * @returns
 */
const delay = function (ms = 1000) {
    return new Promise((resolve: Function) => setTimeout(() => resolve(), ms));
};

/**
 *
 *
 * @export
 * @param {PluginOptions} options
 * @param {Koatty} app Koatty or Koa instance
 */
export async function PluginApollo(options: PluginOptions, app: any) {
    const opt = { ...defaultOptions, ...options };

    /**
     * 
     *
     * @param {*} newConfig
     * @returns
     */
    const reFreshConfig = function (newConfig: any) {
        if (helper.isEmpty(newConfig)) {
            return;
        }
        try {
            const typeConfig: any = {};
            for (const n in newConfig) {
                if (n.includes(".")) {
                    const type = n.slice(0, n.indexOf("."));
                    const key = n.slice(n.indexOf(".") + 1);
                    if (type && key) {
                        // tslint:disable-next-line: no-unused-expression
                        !typeConfig[type] && (typeConfig[type] = {});
                        if (helper.isJSONStr(newConfig[n])) {
                            typeConfig[type][key] = JSON.parse(newConfig[n]);
                        } else {
                            typeConfig[type][key] = newConfig[n];
                        }
                    }
                } else {
                    // tslint:disable-next-line: no-unused-expression
                    !typeConfig.config && (typeConfig.config = {});
                    if (helper.isJSONStr(newConfig[n])) {
                        typeConfig.config[n] = JSON.parse(newConfig[n]);
                    } else {
                        typeConfig.config[n] = newConfig[n];
                    }
                }
            }
            let appConfigs = {};
            if (app.setMap) {
                appConfigs = app.getMap("configs") || {};
                appConfigs = helper.extend(appConfigs, typeConfig, true);
                app.setMap("configs", appConfigs);
            }
        } catch (err) {
            logger.Error(err.stack || err);
        }
    };

    /**
     *
     *
     * @param {*} opt
     * @param {*} app
     */
    const initApollo = async function (opt: any, app: any) {
        // Instantiate Apollo
        const apolloClient = apollo(opt).namespace(opt.namespace, 'JSON');
        if (opt.enableUpdateNotification) {
            apolloClient.on('change', ({ key, oldValue, newValue }: UpdateNotification) => {
                logger.Info(`Apollo configuration key ${key} has been refreshed.`);
                reFreshConfig({ [key]: newValue });
            });
        }

        const ns = await apolloClient.ready().catch((err: Error) => {
            return Promise.reject(err);
        });

        const apolloConfigs = await ns.config() || {};
        if (!helper.isEmpty(apolloConfigs)) {
            reFreshConfig(apolloConfigs);
        }
        logger.Info('Apollo configuration center initialization is complete.');
        helper.define(app, "apolloClient", apolloClient, true);


        return apolloClient;
    };
    helper.define(app, "initApollo", initApollo, true);

    const client = await initApollo(opt, app).catch((err: any) => {
        logger.Error("Apollo configuration center initialization error.", err);
        // app.emit("apolloRetry");
    });

    // retry
    client.on("fetch-error", async function () {
        // await delay(5000);
        return initApollo(opt, app);
    });
}